sqliteparserutils/utils.go

```diff
@@ -14,105 +14,110 @@ type SplitStatementExtraInfo struct {
+	if iterator.currentToken != nil {
+		lastTokenType = iterator.currentToken.GetTokenType()
 	}
-
-	stmts = make([]string, 0)
-	for _, stmtInterval := range stmtIntervals {
-		stmts = append(stmts, tokenStream.GetTextFromInterval(stmtInterval))
+	extraInfo = SplitStatementExtraInfo{
+		IncompleteCreateTriggerStatement: insideCreateTriggerStmt,
+		IncompleteMultilineComment:       insideMultilineComment,
+		LastTokenType:                    lastTokenType,
 	}
-
-	lastTokenType := antlr.TokenInvalidType
-	if previousToken != nil {
-		lastTokenType = previousToken.GetTokenType()
+	statement = ""
+	if startPosition != -1 {
+		statement = iterator.tokenizer.source.GetInputStream().GetText(startPosition, previousToken.GetStop())
 	}
-	return stmts, SplitStatementExtraInfo{IncompleteCreateTriggerStatement: insideCreateTriggerStmt, IncompleteMultilineComment: insideMultilineComment, LastTokenType: lastTokenType}
+	return statement, extraInfo, iterator.tokenizer.IsEOF() || insideMultilineComment
 }
 
-func atCreateTriggerStart(tokenStream antlr.TokenStream) bool {
-	if tokenStream.LT(1).GetTokenType() != sqliteparser.SQLiteLexerCREATE_ {
-		return false
+func SplitStatement(statement string) (statements []string, extraInfo SplitStatementExtraInfo) {
+	iterator := createStatementIterator(statement)
+
+	statements = make([]string, 0)
+	for {
+		statement, extraInfo, isEOF := iterator.Next()
+		if statement != "" {
+			statements = append(statements, statement)
+		}
+		if isEOF {
+			return statements, extraInfo
+		}
 	}
+}
 
-	if tokenStream.LT(2).GetTokenType() == sqliteparser.SQLiteLexerTRIGGER_ {
+func atCreateTriggerStart(tokenStream *bufferedTokenizer) bool {
+	token1 := tokenStream.Get(0).GetTokenType()
+	token2 := tokenStream.Get(1).GetTokenType()
+	token3 := tokenStream.Get(2).GetTokenType()
+
+	if token1 == sqliteparser.SQLiteLexerCREATE_ && token2 == sqliteparser.SQLiteLexerTRIGGER_ {
 		return true
 	}
-
-	if tokenStream.LT(2).GetTokenType() == sqliteparser.SQLiteLexerTEMP_ || tokenStream.LT(2).GetTokenType() == sqliteparser.SQLiteLexerTEMPORARY_ &&
-		tokenStream.LT(3).GetTokenType() == sqliteparser.SQLiteLexerTRIGGER_ {
+	if token1 == sqliteparser.SQLiteLexerCREATE_ &&
+		(token2 == sqliteparser.SQLiteLexerTEMP_ || token2 == sqliteparser.SQLiteLexerTEMPORARY_) &&
+		token3 == sqliteparser.SQLiteLexerTRIGGER_ {
 		return true
 	}
-
 	return false
-
 }
 
 // Note: Only starts for incomplete multiline comments will be detected cause lexer automatically ignores complete
 // multiline comments
-func atIncompleteMultilineCommentStart(tokenStream antlr.TokenStream) bool {
-	if tokenStream.LT(1).GetTokenType() != sqliteparser.SQLiteLexerDIV {
-		return false
-	}
-
-	if tokenStream.LT(2).GetTokenType() == sqliteparser.SQLiteLexerSTAR {
-		return true
-	}
-
-	return false
+func atIncompleteMultilineCommentStart(stream *bufferedTokenizer) bool {
+	token1 := stream.Get(0).GetTokenType()
+	token2 := stream.Get(1).GetTokenType()
```

Comment:
Just to make sure - is this intentional that previously the function was using indexes 1 and 2 and now it uses 0 and 1? I know it accesses different stream but would like to double check there's no off by one here.

Reply:
Yes, `BufferedTokenizer` have 0-based indexing (in a sense) compared to the `CommonTokenStream`